rpc: deprecate `addresses` and `reqSigs` from rpc outputs

mjdietzx opened the pull request:
Considering the limited applicability of `reqSigs` and the confusing output of `1` in all cases except bare multisig, the `addresses` and `reqSigs` outputs are removed for all rpc commands.
    
1) add a new sane "address" field (for outputs that have an identifiable address, which doesn't include bare multisig)
2) with -deprecatedrpc: leave "reqSigs" and "addresses" intact (with all weird/wrong behavior they have now)
3) without -deprecatedrpc: drop "reqSigs" and "addresses" entirely always.
    
Note: Some light refactoring done to allow us to very easily delete a few chunks of code (marked with TODOs) when we remove this deprecated behavior.

Using `IsDeprecatedRPCEnabled` in core_write.cpp caused some circular dependencies involving core_io

Circular dependencies were caused by rpc/util unnecessarily importing node/coinstats and node/transaction. Really what rpc/util needs are some fundamental type/helper-function definitions. So this was cleaned up to make more sense.

This fixes #20102.

jonatack commented on src/rpc/rawtransaction.cpp:
```diff
@@ -560,7 +557,6 @@ static RPCHelpMan decodescript()
                             {RPCResult::Type::STR, "asm", "String representation of the script public key"},
                             {RPCResult::Type::STR_HEX, "hex", "Hex string of the script public key"},
                             {RPCResult::Type::STR, "type", "The type of the script public key (e.g. witness_v0_keyhash or witness_v0_scripthash)"},
-                            {RPCResult::Type::NUM, "reqSigs", "The required signatures (always 1)"},
```
ISTM these RPC removals would need to go through the usual RPC deprecation process?

mjdietzx commented on test/util/data/txcreatemultisig1.json:
```diff
@@ -15,7 +15,6 @@
             "scriptPubKey": {
                 "asm": "2 02a5613bd857b7048924264d1e70e08fb2a7e6527d32b7ab1bb993ac59964ff397 021ac43c7ff740014c3b33737ede99c967e4764553d1b2b83db77c83b8715fa72d 02df2089105c77f266fa11a9d33f05c735234075f2e8780824c6b709415f9fb485 3 OP_CHECKMULTISIG",
                 "hex": "522102a5613bd857b7048924264d1e70e08fb2a7e6527d32b7ab1bb993ac59964ff39721021ac43c7ff740014c3b33737ede99c967e4764553d1b2b83db77c83b8715fa72d2102df2089105c77f266fa11a9d33f05c735234075f2e8780824c6b709415f9fb48553ae",
-                "reqSigs": 2,
```
Based on the discussion in the issue #20102 it seemed no one cares/uses bare multisig, and that we could just totally get rid of `reqSigs`. I'm not sure if this is really true, but figured I'd remove it and see what feedback I get in the PR

MarcoFalke commented:
Is anyone using this for p2sh-multisig (passing the redeemscript into `decodescript`)?

luke-jr commented:
"addresses" for multisigs don't make sense without "reqSigs"

IMO if we deprecate/remove the latter, we should also be sure to remove the former with it.

("addresses" should remain for actual address outputs)

mjdietzx commented:
> "addresses" for multisigs don't make sense without "reqSigs"
> 
> IMO if we deprecate/remove the latter, we should also be sure to remove the former with it.
> 
> ("addresses" should remain for actual address outputs)

If we were to remove `addresses` for multisigs it seems to me we'd be able to get rid of [ExtractDestinations](https://github.com/bitcoin/bitcoin/blob/0ede354367c478cbbed0ec40a9d3cf1efe74cf33/src/script/standard.h#L253) entirely, as this is the only thing it's used for currently, and `ExtractDestination` can be used instead.

It also makes me think ```{RPCResult::Type::ARR, "addresses", "array of bitcoin addresses"}``` is a bit misleading, and would ideally be a single ```{RPCResult::Type::STR, "address", "bitcoin address"}```.

We'd be able to get rid of a nice chunk of code (and potentially confusing behavior). But I'm new to this repo, and I need some guidance on what should be done here. And also if I need to deprecate stuff and how I should proceed if so.

Note: there is always the easy option of just fixing the `reqSigs` behavior instead of removing it. But, I guess it's unused and we can get rid of some cruft by removing it. Please advise!

sipa commented:
Concept ACK on removing both "reqSigs" and "addresses", but it'll need to go through the RPC deprecation mechanism.

In other places we have already removed similar fields. Especially "addresses" is a bizarre and misleading leftover from a time when addresses were only ever P2PKH and just "key identifiers". Now (... since P2SH in 2012) addresses are encodings of a specific scriptPubKey it makes no sense whatsoever that multisig is treated as "multiple addresses".

luke-jr commented:
>Note: there is always the easy option of just fixing the reqSigs behavior instead of removing it. 

If it's broken, we should fix it (for backporting) before removing.

>It also makes me think {RPCResult::Type::ARR, "addresses", "array of bitcoin addresses"} is a bit misleading, and would ideally be a single {RPCResult::Type::STR, "address", "bitcoin address"}.

I'm not sure the array format is entirely/hypothetically useless. Might as well leave it alone (or at least propose this change separately).

mjdietzx commented:
I've re-worked this PR, and force pushed, based on the convo/feedback in this thread. Broken into two commits:

1. https://github.com/bitcoin/bitcoin/pull/20286/commits/f5d3cf6ceb3bb578bfce36959744ee7035eaf6c0 "fixes" `reqSigs`. Instead of defaulting to `1` which is confusing (and possibly incorrect?) `reqSigs` is not included in rpc outputs unless the transaction is a bare multisig.
2. https://github.com/bitcoin/bitcoin/pull/20286/commits/c0545cc866a69da7ee30bf7cd5b9a84262da5778 deprecates `reqSigs` as discussed, where `reqSigs` is totally removed from rpc outputs, and `addresses` is no longer returned for bare multsigs. https://github.com/bitcoin/bitcoin/pull/20286/commits/c0545cc866a69da7ee30bf7cd5b9a84262da5778#diff-80ada085034954b397668e082d78cd40519aa9b162ec52101651dd48c48a17b3L18-L24 basically becomes a functional test, where we check the default case, and `-deprecatedrpc=reqSigs` are both correct

Note: This PR turned out to be more involved than I expected. I'll probably need some more feedback and pointers to get this across the finish line.

Also: we'll be able to delete a few chunks of code when we can get rid of `-deprecatedrpc=reqSigs`. I indicate this in the source with `TODO` comments.

sipa commented:
@mjdietzx I think it's a bit weird to fix `"addresses"`/`"reqSigs"`, as it's making a semantics change simultaneously with a deprecation.

My suggestion would be to:
* add a new sane `"address"` field (for outputs that have an identifiable address, which doesn't include bare multisig)
* with -deprecatedrpc: leave `"reqSigs"` and `"addresses"` intact (with all weird/wrong behavior they have now)
* without -deprecatedrpc: drop `"reqSigs"` and `"addresses"` entirely, always.

That means that anyone who is currently relying on the weird `"reqSigs": `` behavior for unknown address types (for who knows what reason) isn't affected if they set the deprecation option.

mjdietzx commented:
I like the approach you've outlined @sipa. Any objections before I go ahead and do this @luke-jr? I know you weren't quite convinced re: `addresses` => `address`.

luke-jr commented:
I'd prefer to keep a single-item array (better compatibility, and future-proof), but I don't care enough to object.

mjdietzx commented:
I'll need to re-work some of the functional tests (will get this committed soon) now that `addresses` is removed by default, but https://github.com/bitcoin/bitcoin/pull/20286/commits/162e7f0969bf593e66ec2e59e24c4c555cd101f7 updates this PR to be as @sipa suggested

mjdietzx commented:
Piggy backing off of https://github.com/bitcoin/bitcoin/pull/20211, now that we have:

```c++
bool ExtractDestination(const CScript& scriptPubKey, TxoutType& typeRet, CTxDestination& addressRet);
```

Could we do something like:
```c++
TxoutType type;
CTxDestination dest;
ExtractDestination(m_script, type, dest);
switch (type) {
case TxoutType::PUBKEYHASH:
case TxoutType::SCRIPTHASH: return OutputType::LEGACY;
case TxoutType::WITNESS_V0_SCRIPTHASH:
case TxoutType::WITNESS_V0_KEYHASH:
case TxoutType::WITNESS_V1_TAPROOT:
case TxoutType::WITNESS_UNKNOWN: return OutputType::BECH32;
case TxoutType::PUBKEY:
case TxoutType::MULTISIG:
case TxoutType::NONSTANDARD:
case TxoutType::NULL_DATA: return nullopt; // no default case, so the compiler can warn about missing cases
}
assert(false);
```

Instead of what we have now:
```c++
CTxDestination dest;
ExtractDestination(m_script, dest);
switch (dest.which()) {
  case 1 /* PKHash */:
  case 2 /* ScriptHash */: return OutputType::LEGACY;
  case 3 /* WitnessV0ScriptHash */:
  case 4 /* WitnessV0KeyHash */:
  case 5 /* WitnessUnknown */: return OutputType::BECH32;
  case 0 /* CNoDestination */:
  default: return nullopt;
}
```

Note: I messed around with this, and probably made some mistakes here, but didn't see any unit tests or functional tests fail. So maybe some test coverage is needed here as well?

mjdietzx commented:
This PR is ready for review again. Squashed into two commits (bc https://github.com/bitcoin/bitcoin/pull/20286/commits/07ff0f7d15b99a40ec09f1315bcfcd9e4a6099a2 refactor was required to fix a circular dependency lint error introduced by using `IsDeprecatedRPCEnabled()` in `ScriptPubKeyToUniv()`)

mjdietzx commented:
Hey, this PR seems a bit dead in the water. Does anyone have suggestions on how I can help move this through the review process?

My guess is this change is making it hard to review: 
https://github.com/bitcoin/bitcoin/pull/20286/files#diff-d3f64e14005fbfea3d4f72b076764ac897d0df451de0ab2fb1c57a5f87cd793bR23-R26

But that is the most straightforward way I could see to get rid of a circular dependency and allow me to use `IsDeprecatedRPCEnabled` in `core_write.cpp` here: https://github.com/bitcoin/bitcoin/pull/20286/files#diff-48fa743e653b0ceebb1e454766c5a2e70981a44503eca6adf24895016db9eb26R164

Any thoughts/suggestions?

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -132,7 +132,7 @@ static RPCHelpMan getrawtransaction()
                                          {RPCResult::Type::STR, "asm", "the asm"},
                                          {RPCResult::Type::STR, "hex", "the hex"},
-                                         {RPCResult::Type::NUM, "reqSigs", "The required sigs"},
+                                         {RPCResult::Type::NUM, "reqSigs", /* optional */ true, "(DEPRECATED, returned only if config option -deprecatedrpc=reqSigs is passed) Number of required signatures (only present for bare multisig"},
```
I don't think it's technically true anymore in the current PR that this is only present for bare multisig (here and elsewhere).

sipa commented on src/rpc/blockchain.cpp:
```diff
@@ -1104,8 +1104,8 @@ static RPCHelpMan gettxout()
                                 {RPCResult::Type::STR_HEX, "hex", ""},
-                                {RPCResult::Type::NUM, "reqSigs", "Number of required signatures"},
-                                {RPCResult::Type::STR_HEX, "type", "The type, eg pubkeyhash"},
+                                {RPCResult::Type::NUM, "reqSigs", /* optional */ true, "(DEPRECATED, returned only if config option -deprecatedrpc=reqSigs is passed) Number of required signatures (only present for bare multisig"},
```
This change seems to be in the wrong commit.

sipa reviewed (commented):
Concept and approach ACK on the first commit. The circular dependency fix seems weird to me as the CoinStatsHashType really belongs together with GetUTXOStates; moving it out makes its purpose very unclear, and makes the code more confusing, not less. The goal of the circular dependency checker is that such dependencies are usually a sign of bad design or hard-to-follow code, and we want to avoid introducing that. But if the cure is worse than the disease, just add an exception.

mjdietzx commented:
> Concept and approach ACK on the first commit. The circular dependency fix seems weird to me as the CoinStatsHashType really belongs together with GetUTXOStates; moving it out makes its purpose very unclear, and makes the code more confusing, not less. The goal of the circular dependency checker is that such dependencies are usually a sign of bad design or hard-to-follow code, and we want to avoid introducing that. But if the cure is worse than the disease, just add an exception.

Ah yeah that's much better 👍. I addressed this, and your other comments in this commit: 811947c0527fa868157352434a5c310609dd5c63

mjdietzx commented:
> 🐙 This pull request conflicts with the target branch and [needs rebase](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#rebasing-changes).
> 
> Want to unsubscribe from rebase notifications on this pull request? Just convert this pull request to a "draft".

Squashed and rebased

MarcoFalke commented on src/Makefile.am:
```diff
@@ -502,12 +502,14 @@ libbitcoin_common_a_SOURCES = \
   script/sign.cpp \
   script/signingprovider.cpp \
   script/standard.cpp \
+  shutdown.cpp \
```
Can you explain what this is doing and why?

Also, the ci is failing

mjdietzx replied:
Since I've added `rpc/server.cpp` to `libbitcoin_common_a_SOURCES`, and `rpc/server.cpp` in turn includes `shutdown.h` and invokes `StartShutdown()` [here],(https://github.com/bitcoin/bitcoin/blob/master/src/rpc/server.cpp#L172) I need to include `shutdown.cpp` in `libbitcoin_common_a_SOURCES`

Otherwise we'll get a compile error:
```sh
Undefined symbols for architecture x86_64:
  "StartShutdown()", referenced from:
      std::__1::__function::__func<stop()::$_3, std::__1::allocator<stop()::$_3>, UniValue (RPCHelpMan const&, JSONRPCRequest const&)>::operator()(RPCHelpMan const&, JSONRPCRequest const&) in libbitcoin_common.a(libbitcoin_common_a-server.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

mjdietzx replied:
Looks like I made a mistake resolving a merge conflict when I rebased to master, which caused CI to fail.

The CI error was:
```sh
Good job! The circular dependency "policy/fees -> txmempool -> validation -> policy/fees" is no longer present.
458Please remove it from EXPECTED_CIRCULAR_DEPENDENCIES in test/lint/lint-circular-dependencies.sh
459to make sure this circular dependency is not accidentally reintroduced.
```

I had a single merge conflict in `lint-circular-dependencies.sh` where: `"policy/fees -> txmempool -> validation -> policy/fees"` needed to be removed, but I re-added it instead of removing it when resolving the merge conflict.

Fixed now: https://github.com/bitcoin/bitcoin/pull/20286/files#diff-ee6a5803b78eb12f095b391894014b5fb2b836cef29943980b3ab684861c1f2dR23

mjdietzx commented:
One thing I noticed, which I fixed in 2nd commit https://github.com/bitcoin/bitcoin/pull/20286/commits/3c54db1ae676a7c12e09336b793a601c20810bf7, that's a little confusing to me. CI was failing for `[depends, sanitizers: thread (TSan), no gui] [focal]` and `[no depends, sanitizers: address/leak (ASan + LSan) + undefined(UBSan) + integer] [focal]`. You can see the results of that run [here](https://cirrus-ci.com/build/6325232661168128)

It's because I generated a block on `self.nodes[0]`, but didn't `self.sync_all()` before getting that block on `self.nodes[1]`. This all happens in the functional test I added [here](https://github.com/bitcoin/bitcoin/pull/20286/files#diff-fa7d372f6d77cd74c3ca58e0bd66028d1299a72fcdc375ad07ff4216ca634f92R48)

Which resulted in this error (for those two CI runs):
```sh
Traceback (most recent call last):
  ...
  ..
  File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/rpc_reqsigs_deprecation.py", line 50, in test_reqsigs_deprecation
    script_pub_key = self.nodes[1].getblock(blockhash=hash, verbose=2)['tx'][-1]['vout'][0]['scriptPubKey']
  ...
  ..
test_framework.authproxy.JSONRPCException: Block not found on disk (-1)
```

Although it seems correct to call `self.sync_all()` as I do now in the 2nd commit, I'm confused because it didn't fail for most of the other CI runs, and didn't fail when I ran the functional tests locally (Mac). I'm just wondering if anyone knows or can explain why there is different behavior here

mjdietzx commented:
Looks like this ones gonna miss the v0.21.0 boat 🚢

Either way, want to see if there's anything I can do to help move this review forward. All feedback in this review until now has been addressed, and I squashed into single commit

MarcoFalke commented on src/core_write.cpp:
```diff
@@ -141,40 +142,41 @@ void ScriptToUniv(const CScript& script, UniValue& out, bool include_address)
+    if (ExtractDestination(script, type, address) && include_address && type != TxoutType::PUBKEY) {
         out.pushKV("address", EncodeDestination(address));
     }
+    out.pushKV("type", GetTxnOutputType(type));
```
Why are you changing the order?

sipa replied:
This doesn't seem addressed: https://github.com/bitcoin/bitcoin/pull/20286/files#r547134259 (it's ok if you disagree with the comment, or there is a good reason to - but do respond if you can).

mjdietzx replied:
The first thing done in `ExtractDestination` is
```c++
std::vector<valtype> vSolutions;
whichType = Solver(scriptPubKey, vSolutions);
```
so prior to this change, `Solver` code was duplicated and running twice.

By changing the order in the if statement from `include_address && ExtractDestination(script, address` => `ExtractDestination(script, type, address) && include_address` we are guaranteed to always have `type` set by `ExtractDestination`, and can get rid of the duplicated `Solver` code here.

MarcoFalke commented on src/core_write.cpp:
```diff
@@ -141,40 +142,41 @@ void ScriptToUniv(const CScript& script, UniValue& out, bool include_address)
-        out.pushKV("type", GetTxnOutputType(type));
-        return;
+    // TODO: from v0.22 ("addresses" and "reqSigs" deprecated) this entire if statement should be removed
+    if (IsDeprecatedRPCEnabled("reqSigs")) { // cruft...
```
Seems like a layer violation. You can add a `ScriptPubKeyToUniv` to `rpc/blockchain.h`, which calls this `ScriptPubKeyToUniv` and passes an additional boolean parameter with the value of `IsDeprecatedRPCEnabled('reqsiqs')`

MarcoFalke commented on src/core_write.cpp:
```diff
@@ -141,40 +142,41 @@ void ScriptToUniv(const CScript& script, UniValue& out, bool include_address)
+        out.pushKV("address", EncodeDestination(address));
     }
-    out.pushKV("addresses", a);
+    out.pushKV("type", GetTxnOutputType(type));
```
Same

mjdietzx replied:
I changed the order here so that when the deprecation period is over, we just simply delete the `if (fIncludeDeprecated) {` branch. It also cleaned things up a bit as we don't need to push `out.pushKV("type", GetTxnOutputType(type));` twice. The behavior is the same though

mjdietzx replied:
Anyways, if you think this is a bad idea or doesn't belong in this PR, I have no problem undoing these changes and trying to stick to the original code as closely as possible - just lmk. Part of the reason I added some fuzz tests in this PR https://github.com/bitcoin/bitcoin/pull/20772 was to help build confidence since these two functions have to be touched in this PR either way

MarcoFalke commented on src/script/standard.cpp:
```diff
@@ -221,6 +221,12 @@ bool ExtractDestination(const CScript& scriptPubKey, CTxDestination& addressRet)
     return false;
 }
 
+bool ExtractDestination(const CScript& scriptPubKey, CTxDestination& addressRet) {
```
```suggestion
bool ExtractDestination(const CScript& scriptPubKey, CTxDestination& addressRet)
{
```

maflcko reviewed (commented):
I still don't like changing the build system for a minor rpc change

sipa commented on src/core_write.cpp:
```diff
@@ -141,40 +141,42 @@ void ScriptToUniv(const CScript& script, UniValue& out, bool include_address)
-        return;
+    // TODO: from v0.22 ("addresses" and "reqSigs" deprecated) this entire if statement should be removed
+    const std::vector<std::string> enabled_methods = gArgs.GetArgs("-deprecatedrpc");
+    if (find(enabled_methods.begin(), enabled_methods.end(), "reqSigs") != enabled_methods.end()) {
```
I'd think that "addresses" being removed is a more prominent change, so perhaps the deprecatedrpc name should be "addressed" ?

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -132,7 +132,7 @@ static RPCHelpMan getrawtransaction()
-                                         {RPCResult::Type::NUM, "reqSigs", "The required sigs"},
+                                         {RPCResult::Type::NUM, "reqSigs", /* optional */ true, "(DEPRECATED, returned only if config option -deprecatedrpc=reqSigs is passed) Number of required signatures"},
                                          {RPCResult::Type::STR, "type", "The type, eg 'pubkeyhash'"},
                                          {RPCResult::Type::ARR, "addresses", "",
```
This should probably also get a changed comment.

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -490,7 +490,7 @@ static RPCHelpMan decoderawtransaction()
-                                    {RPCResult::Type::NUM, "reqSigs", "The required sigs"},
+                                    {RPCResult::Type::NUM, "reqSigs", /* optional */ true, "(DEPRECATED, returned only if config option -deprecatedrpc=reqSigs is passed) Number of required signatures"},
                                     {RPCResult::Type::STR, "type", "The type, eg 'pubkeyhash'"},
                                     {RPCResult::Type::ARR, "addresses", "",
```
Same here.

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -548,7 +548,7 @@ static RPCHelpMan decodescript()
                         {RPCResult::Type::STR, "type", "The output type (e.g. "+GetAllOutputTypes()+")"},
-                        {RPCResult::Type::NUM, "reqSigs", "The required signatures"},
+                        {RPCResult::Type::NUM, "reqSigs", /* optional */ true, "(DEPRECATED, returned only if config option -deprecatedrpc=reqSigs is passed) Number of required signatures"},
                         {RPCResult::Type::ARR, "addresses", "",
```
And here.

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -559,7 +559,6 @@ static RPCHelpMan decodescript()
                             {RPCResult::Type::STR_HEX, "hex", "Hex string of the script public key"},
                             {RPCResult::Type::STR, "type", "The type of the script public key (e.g. witness_v0_keyhash or witness_v0_scripthash)"},
-                            {RPCResult::Type::NUM, "reqSigs", "The required signatures (always 1)"},
                             {RPCResult::Type::ARR, "addresses", "(always length 1)",
```
And here.

sipa commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3672,9 +3673,9 @@ class DescribeWalletAddressVisitor : public boost::static_visitor<UniValue>
-        } else if (which_type == TxoutType::MULTISIG) {
-            // Also report some information on multisig scripts (which do not have a corresponding address).
-            // TODO: abstract out the common functionality between this logic and ExtractDestinations.
+        } else if (find(enabled_methods.begin(), enabled_methods.end(), "reqSigs") != enabled_methods.end() &&
```
It's pretty ugly to duplicate this logic in many places. I'd suggest just using `IsDeprecatedRPCEnabled`, in the top-level RPC code, and pass down a boolean to wherever it's needed (e.g. in core_write, make `ScriptPubKeyToUniv` take a boolean to indicate whether to not to include `reqSigs` / `addresses`).

mjdietzx replied:
I was able to do this how you and @MarcoFalke suggested in this commit https://github.com/bitcoin/bitcoin/pull/20286/commits/dba1053a5e0c74e53ded3d42c3826373474ba805. However, I'm a little confused by what's going on in `rpcwallet.cpp`. I'm able to use `IsDeprecatedRPCEnabled` here when I build and test locally (macOS), and it seems some of the CI test runs pass. However, some of the CI runs throw a compile error in this case. Any ideas or suggestions on the approach I should take for `rpcwallet`?

sipa commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3672,9 +3673,9 @@ class DescribeWalletAddressVisitor : public boost::static_visitor<UniValue>
-            // TODO: abstract out the common functionality between this logic and ExtractDestinations.
+        } else if (find(enabled_methods.begin(), enabled_methods.end(), "reqSigs") != enabled_methods.end() &&
+                   which_type == TxoutType::MULTISIG) {
+            // TODO: from v0.22 ("addresses" and "reqSigs" deprecated) this entire else if statement should be removed
```
Note that the next major release after 0.21 is going to be v22 (not v0.22); see https://github.com/bitcoin/bitcoin/pull/20223.

mjdietzx commented on test/util/data/txcreatemultisig1.json:
```diff
@@ -15,13 +15,8 @@
-                    "1FXtz9KU8JNmQDyHdiEm5HDiALuP3zdHvV",
-                    "14LuavcBbXZYJ6Tsz3cAUQj9SuQoL2xCQX"
-                ]
+                "address": "1FoG2386FG2tAJS9acMuiDsKy67aGg9MKz",
```
Please note this behavior for bare multisigs. I'm not totally sure what we want `address` to be in this case. For now I just made it the first `address` that `ExtractDestinations` sets in `addresses`. But maybe `address` should not be included for `type == TxoutType::MULTISIG`?

Looking forward to when we get rid of the `deprecatedrpc=addresses` logic, we'll end up removing `ExtractDestinations` completely. And at that point we won't necessarily have an `addresses` array for `type == TxoutType::MULTISIG` - so not including `address` in the response for bare multisigs would seem to be the long-term (and maybe more correct) solution?

sipa replied:
Bare multisig outputs should not have any address, because there is no address that corresponds to their scriptPubKey.

mjdietzx replied:
Ok, done https://github.com/bitcoin/bitcoin/pull/20286/commits/98b7080d7eeece1081eed3c68dd092e974ad37ee

sipa commented on src/rpc/blockchain.cpp:
```diff
@@ -1778,6 +1778,11 @@ void CalculatePercentilesByWeight(CAmount result[NUM_GETBLOCKSTATS_PERCENTILES],
     }
 }
 
+void ScriptPubKeyToUniv(const CScript& scriptPubKey, UniValue& out, bool fIncludeHex)
```
There is only one invocation of this function in this .cpp file. Maybe just inline it?

sipa commented on src/rpc/net.cpp:
```diff
@@ -225,12 +225,12 @@ static RPCHelpMan getpeerinfo()
         obj.pushKV("bip152_hb_from", stats.m_bip152_highbandwidth_from);
         if (IsDeprecatedRPCEnabled("getpeerinfo_addnode")) {
-            // addnode is deprecated in v0.21 for removal in v0.22
+            // addnode is deprecated in v0.21 for removal in v22
```
Nit: no need to change unrelated code. I just mentioned the version numbering change as an FYI.

sipa replied:
Actually, this is incorrect. Current master will become v22. These things can be removed in v23.

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -6,7 +6,6 @@
 #include <chain.h>
 #include <coins.h>
 #include <consensus/validation.h>
-#include <core_io.h>
```
This file is still used.

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -132,9 +130,10 @@ static RPCHelpMan getrawtransaction()
+                                         {RPCResult::Type::NUM, "reqSigs", /* optional */ true, "(DEPRECATED, returned only if config option -deprecatedrpc=addresses is passed) Number of required signatures"},
                                          {RPCResult::Type::STR, "type", "The type, eg 'pubkeyhash'"},
-                                         {RPCResult::Type::ARR, "addresses", "",
+                                         {RPCResult::Type::STR, "address", /* optional */ true, "bitcoin address (not included for bare multisig)"},
```
Same comment about mentioning bare multisig here being confusing.

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -490,9 +489,10 @@ static RPCHelpMan decoderawtransaction()
+                                    {RPCResult::Type::NUM, "reqSigs", /* optional */ true, "(DEPRECATED, returned only if config option -deprecatedrpc=addresses is passed) Number of required signatures"},
                                     {RPCResult::Type::STR, "type", "The type, eg 'pubkeyhash'"},
-                                    {RPCResult::Type::ARR, "addresses", "",
+                                    {RPCResult::Type::STR, "address", /* optional */ true, "bitcoin address (not included for bare multisig)"},
```
And here.

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -548,8 +548,9 @@ static RPCHelpMan decodescript()
                         {RPCResult::Type::STR, "type", "The output type (e.g. "+GetAllOutputTypes()+")"},
-                        {RPCResult::Type::NUM, "reqSigs", "The required signatures"},
-                        {RPCResult::Type::ARR, "addresses", "",
+                        {RPCResult::Type::STR, "address", /* optional */ true, "bitcoin address (not included for bare multisig)"},
```
And here.

sipa commented on src/test/fuzz/script.cpp:
```diff
@@ -4,10 +4,10 @@
 
 #include <chainparams.h>
 #include <compressor.h>
-#include <core_io.h>
```
Still used.

sipa commented on src/test/fuzz/script.cpp:
```diff
@@ -4,10 +4,10 @@
 #include <core_memusage.h>
 #include <policy/policy.h>
 #include <pubkey.h>
+#include <rpc/blockchain.h>
```
This is pretty ugly. The fuzz test shouldn't be using command-line options at all; just invoke the explicit-addresses version once with false and once with true, if you need to.

sipa commented:
The compile error is probably due to a library dependency order (wallet/rpcwallet and rpc/server are in different libraries, so they may not be able to use each other in both directions). Try using `pwallet->chain().rpcEnableDeprecated("addresses")`.

sipa commented on src/script/standard.cpp:
```diff
@@ -221,6 +221,7 @@ bool ExtractDestination(const CScript& scriptPubKey, CTxDestination& addressRet)
     return false;
 }
 
+// TODO: from v22 ("addresses" and "reqSigs" deprecated) "ExtractDestinations" should be removed
```
This is incorrect. Current master will become v22. This functionality can be removed in v23.

MarcoFalke commented:
Please squash your commits according to https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits

sipa commented on src/core_write.cpp:
```diff
@@ -169,17 +172,22 @@ void ScriptPubKeyToUniv(const CScript& scriptPubKey,
     }
 
-    out.pushKV("reqSigs", nRequired);
+    if (ExtractDestination(scriptPubKey, address)) {
```
See my earlier edit here: https://github.com/bitcoin/bitcoin/pull/20286#discussion_r549429414

For "address" we shouldn't report anything for P2PK outputs either, as the address that's reported currently is really the P2PKH one for the same pubkey (and sending to that P2PKH address would have a distinct scriptPubKey). This is due to historical reasons where addresses used to also be "pubkey id" (that's the same reason as why "addresses" were reported for bare multisig; those addresses in modern meaning have nothing to do with that multisig address).

Can you add a `&& GetScriptForDestination(address) == scriptPubKey`? That would guarantee that whatever address is extracted exactly matches the output.

mjdietzx replied:
I pushed a commit as you suggest: https://github.com/bitcoin/bitcoin/pull/20286/commits/97fbaffc4aea265d638895531dc3b3c84669c561

I'm a little confused though. Aren't we already not reporting anything for P2PK bc we return here https://github.com/bitcoin/bitcoin/blob/master/src/core_write.cpp#L167, if `type == TxoutType::PUBKEY`?

Also, I want to know if this problem could affect us anywhere else. It seems `ExtractDestination` already handles this https://github.com/bitcoin/bitcoin/blob/master/src/core_write.cpp#L151 (only pushes an `address` if `type != TxoutType::PUBKEY`) -- but if not, `rawtransaction.cpp` invokes `ScriptToUniv` which may need this check as well

Last, I'm trying to figure out if anything needs to be done in `rpcwallet`, here: https://github.com/bitcoin/bitcoin/blob/master/src/wallet/rpcwallet.cpp#L3669

sipa replied:
> I'm a little confused though. Aren't we already not reporting anything for P2PK bc we return here https://github.com/bitcoin/bitcoin/blob/master/src/core_write.cpp#L167, if `type == TxoutType::PUBKEY`?

You're right, this change isn't needed. It seems the specific P2PK case was already addressed earlier.

> Also, I want to know if this problem could affect us anywhere else. It seems `ExtractDestination` already handles this https://github.com/bitcoin/bitcoin/blob/master/src/core_write.cpp#L151 (only pushes an `address` if `type != TxoutType::PUBKEY`) -- but if not, `rawtransaction.cpp` invokes `ScriptToUniv` which may need this check as well

The code you're linking to is `ScriptToUniv`, not `ExtractDestination`. `ExtractDestination` does produce output for P2PK outputs. It makes sense to see what other call sites may be relying on this P2PK behavior though; if there aren't any, or they are all bogus, it may be easier to just drop support for P2PK in `ExtractDestinations`.

> Last, I'm trying to figure out if anything needs to be done in `rpcwallet`, here: https://github.com/bitcoin/bitcoin/blob/master/src/wallet/rpcwallet.cpp#L3669

That looks like a bug, but a hard to trigger one. You'd need a P2PK script inside a P2SH or P2WSH script in a wallet for that.

mjdietzx replied:
OK, I'll go ahead and revert the commit since it isn't needed.

> The code you're linking to is `ScriptToUniv`, not `ExtractDestination`. `ExtractDestination` does produce output for P2PK outputs. It makes sense to see what other call sites may be relying on this P2PK behavior though; if there aren't any, or they are all bogus, it may be easier to just drop support for P2PK in `ExtractDestinations`.

Ah yeah good point. Well I'll read through the code and try to figure out if an issue needs to be created. I guess if `ExtractDestination(s)` dropped support for `P2PK` it may also fix`rpcwallet` (if it does turn out to be a bug) https://github.com/bitcoin/bitcoin/blob/master/src/wallet/rpcwallet.cpp#L3669. It'll take me some time to fully understand all this and test some things out

MarcoFalke commented:
Btw, you don't have to mention me in the commit message. All I did was some minor review suggestion. :sweat_smile:

mjdietzx commented:
> Btw, you don't have to mention me in the commit message. All I did was some minor review suggestion. 😅

Ah ok, I think GH automatically added that since I committed your suggestion, I just took it out.

Anyways thank you @MarcoFalke @sipa for all the review so far. I know this has been a kinda brutal review process for a minor change - I'm learning so write it off as an investment in a future contributor 😬

sipa commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3671,9 +3672,8 @@ class DescribeWalletAddressVisitor : public boost::static_visitor<UniValue>
-        } else if (which_type == TxoutType::MULTISIG) {
-            // Also report some information on multisig scripts (which do not have a corresponding address).
-            // TODO: abstract out the common functionality between this logic and ExtractDestinations.
+        } else if (which_type == TxoutType::MULTISIG && pwallet->chain().rpcEnableDeprecated("addresses")) {
```
Just noticed this: please don't remove this. This isn't doing anything confusing (it's reporting actual pubkeys, not addresses), and isn't restricted to bare multisig. Reporting detailed information about properly wrapped multisig isn't something we should remove.

mjdietzx replied:
Thanks for catching that - I reverted this change and updated the docs. re-squashed

sipa commented:
utACK 3c2b24b8f5798976e4d9b6b873f423e955098b47

MarcoFalke commented on src/bitcoin-tx.cpp:
```diff
@@ -725,7 +725,7 @@ static void MutateTx(CMutableTransaction& tx, const std::string& command,
 {
     UniValue entry(UniValue::VOBJ);
-    TxToUniv(tx, uint256(), entry);
+    TxToUniv(tx, uint256(), entry, false); // TODO: bitcoin-tx doesn't go through the addresses deprecation phase currently, defaults to new behavior. Do we want to add a deprecation flag here?
```
e2b814e315f011b6afea26a1ac10aa9f31b5fddb: This needs release notes 

```suggestion
    TxToUniv(tx, uint256(), entry, /* include_addresses */ false);
```

MarcoFalke commented on src/core_io.h:
```diff
@@ -44,8 +44,8 @@ UniValue ValueFromAmount(const CAmount& amount);
+void ScriptPubKeyToUniv(const CScript& scriptPubKey, UniValue& out, bool fIncludeHex, bool include_addresses);
 void ScriptToUniv(const CScript& script, UniValue& out, bool include_address);
-void TxToUniv(const CTransaction& tx, const uint256& hashBlock, UniValue& entry, bool include_hex = true, int serialize_flags = 0, const CTxUndo* txundo = nullptr);
+void TxToUniv(const CTransaction& tx, const uint256& hashBlock, UniValue& entry, bool include_addresses, bool include_hex = true, int serialize_flags = 0, const CTxUndo* txundo = nullptr);
```
e2b814e315f011b6afea26a1ac10aa9f31b5fddb:

This is not safe against silent merge conflicts. Passing an array of bool args (some with default values) without named args can lead to bugs. I think there are two options:

* Remove all default args (might be controversial)
* Change the order so that no two neighbouring args have the same type

I think you can just add the new arg to the beginning (or close to it). It will be removed anyway, so any style concerns that the arg is "too early" are irrelevant.

mjdietzx replied:
I updated `TxToUniv` to use this wrapper approach, so it is now used in both cases. As to why I went with this wrapper approach in the first place, it was suggested earlier in the review (unless I mis-interpreted something):

`I'd suggest just using IsDeprecatedRPCEnabled, in the top-level RPC code, and pass down a boolean to wherever it's needed (e.g. in core_write, make ScriptPubKeyToUniv take a boolean to indicate whether to not to include reqSigs / addresses).`

I know it reduces the diff a bit and meant we didn't have changes in as many files. I *think* it got rid of a circular include dependency as well (since we don't have to include `rpc/server` in a bunch of places now). -- and it may have even been necessary to allow this PR to happen without touching the build system as including `rpc/server` in some files was causing trouble. I don't entirely remember, but know it was done for sane reasons, and I think you had suggested something like this early in the review (although your earlier review comments aren't showing in this PR now)

mjdietzx replied:
I meant for this to be a response to your other comment:
"Can you explain why you are using this wrapper approach for ScriptPubKeyToUniv but not for TxToUniv. Either use it for both or none."

MarcoFalke commented on src/rpc/blockchain.cpp:
```diff
@@ -1120,9 +1120,10 @@ static RPCHelpMan gettxout()
                             {
                                 {RPCResult::Type::STR_HEX, "asm", ""},
```
unrelated nit

```suggestion
                                {RPCResult::Type::STR, "asm", ""},
```

MarcoFalke commented on src/script/standard.cpp:
```diff
@@ -221,6 +221,7 @@ bool ExtractDestination(const CScript& scriptPubKey, CTxDestination& addressRet)
     return false;
```
same commit:

Should probably remove the comment. "Multisig txns have more than one address..."

maflcko reviewed (approved):
review ACK 3c2b24b8f5798976e4d9b6b873f423e955098b47 🛡

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 3c2b24b8f5798976e4d9b6b873f423e955098b47 🛡
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjqnQv7BJdIKCHx584yF7InMc3RLX3f+MwsObnyUTwb6SLgnrJ/1MR/4ge1ksuG
d9lmNZm3sGqZ75q7Go8AyfkVSXqngmyL3tI04wjjpcO7B+Xu1hE+Rc9O4reAabag
yjKsKkrtsT0UkeMYT2eB7pFXdwgfaPjUSDMjuGi9erpzhDTh0ej3PEdD/I/+Q0HH
/FkYcMK4gecdTzPhMiktFp9eWfcBuEq9YEOrpNJK0gRgK9kQdyQb8ui7dYfUhpQw
+t6WHaoYzJwIHxdoY7D29aL5RexblXASgdvD26U2U2FNgbOVTyZIzaHfWXn9l5OK
neU7/7kyDNxAYm4E2RDq+BK/aX0SOC6RjSKR+0RtKcKMA0Tl+Uf9tNfJbW7gdRTd
vHDrfqyBMLJvYY16SteV9p7h22eDeVZ8zogNc+zFTvq03OjKPTTb+4VRdzz/J4yf
7Zqjm03AR6VVumUJJTG7IB/dRTouxJB2cvC8Fr5r0cHeDyxX7lGEsjpSlVMqISuY
oUegAZXN
=yCxp
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `ee289e0483353e4cddc99875c3df748a70e7aeff1a9baf49bc6acdaa7ba688c3  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108ee289e0483353e4cddc99875c3df748a70e7aeff1a9baf49bc6acdaa7ba688c3f010d09ab5adf55d9e92999c324694257fee08fff010367d9a652ae39381f0f33e20aac9f35f08f104600ad937f0080debd746fc5265660083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0105872d3367e8fe96c04bfbcf9e5a51f3608f104600ad938f0083dee70ed1b2a8b1a0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff0106a12f357aeb96c8bd4e3c6d00af3334508f104600ad937f008faedfee4c86055dc0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010f4b74fda7ea3b2d33e7ce7244ea1654208f104600ad938f0080a0f9f96c7ec461b0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

MarcoFalke commented on doc/release-notes.md:
```diff
@@ -69,6 +69,16 @@ Updated RPCs
+  point the deprecation will be removed entirely. Note that these fields are attributes of
+  the `scriptPubKey` object returned in the RPC response. However, in the response
+  of `decodescript` these fields are top-level attributes, and included again as attributes
+  of the `scriptPubKey` object. (#20286)
```
Also, missing release notes for rest

MarcoFalke commented on src/rpc/blockchain.cpp:
```diff
@@ -1778,6 +1779,11 @@ void CalculatePercentilesByWeight(CAmount result[NUM_GETBLOCKSTATS_PERCENTILES],
+void ScriptPubKeyToUniv(const CScript& scriptPubKey, UniValue& out, bool fIncludeHex)
+{
+    ScriptPubKeyToUniv(scriptPubKey, out, fIncludeHex, IsDeprecatedRPCEnabled("addresses"));
+}
```
Can you explain why you are using this wrapper approach for `ScriptPubKeyToUniv` but not for `TxToUniv`. Either use it for both or none.

mjdietzx commented:
Thanks for the review @MarcoFalke I will get all your comments addressed in the next few days (I'm getting through a busy period w my company right now so slightly slow)

mjdietzx commented:
@MarcoFalke your remaining comments are all addressed in this commit 2b74c73. I haven't squashed yet because I'm not sure if that would make reviewers lives more difficult or not. Anyways I'll squash it as appropriate once I get the "please squash" and address any remaining feedback

MarcoFalke commented on src/test/fuzz/transaction.cpp:
```diff
@@ -108,10 +108,10 @@ FUZZ_TARGET_INIT(transaction, initialize_transaction)
-        TxToUniv(tx, /* hashBlock */ {}, u, true);
-        TxToUniv(tx, /* hashBlock */ {}, u, false);
+        TxToUniv(tx, /* hashBlock */ {}, /* include_addresses */ true, u);
+        TxToUniv(tx, /* hashBlock */ {}, false, u);
```
nit: could mention `/* include_addresses */` here as well for symmetry

mjdietzx replied:
squashed, and I included this nit in the push

maflcko reviewed (commented):
Please squash your commits according to https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits

I review from scratch or with range-diff, so squashing will simplify review for me. I think we generally prefer to have a squashed and ready version on every push.

mjdietzx committed:
rpc: deprecate `addresses` and `reqSigs` from rpc outputs

1) add a new sane "address" field (for outputs that have an
   identifiable address, which doesn't include bare multisig)
2) with -deprecatedrpc: leave "reqSigs" and "addresses" intact
   (with all weird/wrong behavior they have now)
3) without -deprecatedrpc: drop "reqSigs" and "addresses" entirely,
   always.

mjdietzx committed:
doc: Add release notes for -deprecatedrpc=addresses and bitcoin-tx

MarcoFalke commented:
re-ACK 732ad2d87bd15db2f42313e0f92b2857f0837f60 only changes are to address feedback 🤗

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK 732ad2d87bd15db2f42313e0f92b2857f0837f60 only changes are to address feedback 🤗
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUin6gv/UeOu6b4VrhEqVZzrvybDcAxYkJ96F6+fuwbmf6b4WrQor81jOCcUvkqy
HfQKi45XS8zF+zfMlvZWc8CrDzluMtvpSxqV94aVwBojVB30jQZkEdxUFlc7Cups
KMtz4QBXqZxNtU4hWy342IYZdAJlndjWJHdOom0JNFFs7mKvrerh3WwGc1VKv0t0
098eTDGZuM/nzmrcy8LriWaYD4knxpBi7VLRHPZHDUWxfM7LFlAtKGHWlkZ0Rwqk
19ornEMPtVERHtFKouKSXHL4cx1qA9lPnUvY6RY653aqeKHtBMglBUoFM5tVOx5n
+k345Z347dhi23hkiN3sYDHITCYA6W7nvRIvPxGxM8+Tndpx/4W5+JHLxOAQwOmH
cLnPQhK1TAl5MQ5k9/8Tb61I1ReXd8UjND2VsY662qtbI8iuaroqnDblnA0xua98
NofIkPZa6S3rQ5Oa1zZWXHEmQKItiqul7E6Fi2SEMiVHQGKSGWLoTO6HIOLWneZn
lR+JWihV
=uo81
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `5b7ea3b2ea13987cc0104b0f945e60e835ec583f2b93a088dc340f7e6587fe4b  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401085b7ea3b2ea13987cc0104b0f945e60e835ec583f2b93a088dc340f7e6587fe4bf0101790309601184e2b61c44a8cc1b5d20808fff010006574f38218caa1db04a03f436eccd708f0209305c7938bc89fea903602b8d56765e393ec9decb2218357ad1baf7d27b4fbd308f020bd77502e0946e0c5affa090644352a4cac6207385aaf30f5088ff85273e60bab08f104602567dff0082069a7fe66092a770083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0105c8fa0ef4ef576f5c5b04e9203a76ba908f1203ec1e30f71be71889ea72b8925870f01709ac8630cfec805c00b761b7c8f372308f104602567def008a41c0a0874a279c80083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01089e5d6ce2eeb9115401c009765e5901908f10460255a0df008300d07aa87efa17f0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010cbc41377ea0cbd4c05ee2d13061fe41408f104602567def008062addb945b682900083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

sipa commented:
Needs rebase.

MarcoFalke commented:
re-ACK 0d83a7326385bf012c1a34d766b96af11fa88979 only change is rebase 🤸

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK 0d83a7326385bf012c1a34d766b96af11fa88979 only change is rebase 🤸
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhYywwAjgK4ihIQ39KVE2lzOU0k1tOoysgSg2R0ecgHzTn7dVymB6zzYYv6EUxX
MSafRTRy9PoTxs7yIijFvI7KeYGMNVfjoz31UvB17ecxPAUetuHR8spLAmmVtdf/
po8kLlJ91r88oUK/of56Mt9jzrIo/UPnjuoznCeRaYgTDo4QQoB/t9l+ZJqA/JCK
Z//4UJDu3cV3fVSa1bnoj+V0aw2GRGt33D7ZFo6+aav3ISbT2E+yMcarBR5Osb3A
nQbXU0DhVb5xv9HOl3tNFQA2i1TllJ63uDSgB+J0ylA9muEmfQJ+6hclkjEzppop
JS3gMNPC1OWU6tj4KCwK/F6tPdSr4sef217IwyQOQpdLU+zn6ojIIPdThPCPBJ0U
pToy32H+zeX4lQJV/3H2st8ylGs4ba2vIHe0Heo6hdOqkUPmk0q/NOBYLK0gRljh
H/y728bGj+aAw4GWh48x0B+3WgRylmjgw+Zk6Zsppg4P6V3nK79lIkgKSKESnzHx
gX/dFts/
=l+z1
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `1e142a9aeeb7dbbc160362a7c3f047702eee9212dda1795b35be97cc096c15f8  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401081e142a9aeeb7dbbc160362a7c3f047702eee9212dda1795b35be97cc096c15f8f010868b3ee964dcd3276f5920ee8cb3705808fff01062157c261c1abf00931ee888c334848a08f10460337b8ff0081c4a3627a52495870083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff0107a7990bc4211469a7eabda32d6b228f508f1209f724fd92c9951b7314b436a0f5b2e975fce2981dac872e3e676b98f2995645708f10460337b8ef00896bf4342410accef0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010d6bbe6f02e71a736804fa80ee16a1c4f08f10460336dbbf008e6eda56834930a340083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010ed79bb8df44ed017a2af5ead4bef46f508f10460337b8ff0083ee76eed47c905820083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

sanket1729 reviewed (approved):
Code review ACK 0d83a7326385bf012c1a34d766b96af11fa88979 . The address feild is indeed a much better RPC output.

MarcoFalke commented:
@sipa Mind to re-ACK? Should be close to merge, hopefully.

MarcoFalke commented:
re-ACK 90ae3d8ca68334ec712d67b21a8d4721c6d79788 📢

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK 90ae3d8ca68334ec712d67b21a8d4721c6d79788 📢
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUg6IAv+Ol7ckNRTmCxL9Ko6DB0jxK62hxEeAODWHiXIrB9y/1wT7WScw8VIBSRn
3p9VIwDE/+WXoU1OUXJcUfszI/vGoWkZO/uO91fQ3ZmN683+55zkh+5t6qIuSXWU
9aqb4EZEhiJtwa1NZOBohZCSdU1c/ogwUN0jn2WNhZwoGR7fccpay5L8K3cLpu3Y
83pCYZ29WrxGeKgq1iV+mW018oh6+/7OvgBEujnM+4qerRDHps/J8fyNBYGuIY8a
iK32n//OmKxP57lY1kbHlUjL5KXW/EWtQiu7srqJsOoi0+ypmCGvDepF3gpbnOVV
OKchPIDkOPBggEQzAF6AeLPUtr+/K6+sZkpImlmo/xvF0KRuwHrUh8f18N/2vZws
zyC5pkSLP+ChVXfW7HnjjmRC3ax5y3bR42XJm5wIXYk4+QLmXMQPqMe+RzxZVr+4
SNaQ8vWQZSSbl+X01TGgyTatrF7dwYtgpLRSImf6C5SXZXB16VLCLoqx+zr3EI3/
gMbBV9Eo
=TB20
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `a41966e5952e2cb0d54de33ddc00b633b1c94ca2c9647f9886491c77a18134ee  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108a41966e5952e2cb0d54de33ddc00b633b1c94ca2c9647f9886491c77a18134eef0102fc2182836732489ef0e0db6ca82a45f08fff0102683974ea938700512d81ccc95f8ae5608f1046059f8b5f008ba34288b9a7fa2580083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010495547f6e8fc8390685d30ee99105b9308f104605a068bf0082fa4c823dc5bac7e0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010af716622f95da00c77c2df0c0218551a08f104605a068bf008952cdc8ddb4305b50083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010c6e2850ec7b24c2ee2487292c1c5490308f104605a068bf0087071ee8a58174de00083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>
